src/middlewares/information.js

```diff
@@ -1,5 +1,15 @@
-module.exports = (req, _, next) => {
+module.exports = (req, res, next) => {
   req.clientIP = req.headers["x-forwarded-for"] || req.connection.remoteAddress;
   req.timestamp = Date.now();
+  req.origin =
+    req.headers.origin ||
+    req.headers.referer ||
+    req.session?.loginAfterState?.redirectOrigin; // sparcssso/callback 요청은 헤더에 origin이 없음
+
+  if (!req.origin) {
+    return res.status(412).json({
+      error: "Precondition Failed : request must have origin in header",
+    });
```

Comment:
우와 저는 HTTP 412 코드를 오늘 처음 알았어요!
HTTP 412 상태 코드는 GET이나 HEAD가 아닌 조건부 요청(POST 요청 등)에서 헤더에 의해 정의된 조건들이 충족되지 않을 때 반환되는 코드래요. 400 Bad request는 어떨까요? 사용자가 임의로 헤더에서 origin을 누락시키는 것은 "deceptive request routing"에 해당할 수 있습니다([400 Bad Request MDN 문서](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400)).
> This happens with conditional requests on methods other than [GET](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET) or [HEAD](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD) when the condition defined by the [If-Unmodified-Since](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Unmodified-Since) or [If-None-Match](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match) headers is not fulfilled.
출처 - https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/412

Reply:
옹 그렇군요. `400`으로 수정했어요 ~